wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceHandler.java

```diff
@@ -0,0 +1,353 @@
+  @Path("v2/contexts/{context}/directives")
+  public void getDirectives(HttpServiceRequest request, HttpServiceResponder responder,
+                            @PathParam("context") String namespace) {
+    respond(responder, namespace, ns -> {
+      // CDAP-15397 - reload must be called before it can be safely used
+      composite.reload(namespace);
+      List<DirectiveUsage> directives = new ArrayList<>();
+      for (DirectiveInfo directive : composite.list(namespace)) {
+        DirectiveUsage directiveUsage = new DirectiveUsage(directive.name(), directive.usage(),
+                                                           directive.description(), directive.scope().name(),
+                                                           directive.definition(), directive.categories());
+        directives.add(directiveUsage);
+      }
+      responder.sendJson(new ServiceResponse<>(directives));
+    });
+  }
+
+  /**
+   * Get the specification for the workspace
+   */
+  @GET
+  @Path("v2/contexts/{context}/workspaces/{id}/specification")
+  public void specification(HttpServiceRequest request, HttpServiceResponder responder,
+                            @PathParam("context") String namespace,
+                            @PathParam("id") String workspaceId) {
+    respond(responder, namespace, ns -> {
+      if (ns.getName().equalsIgnoreCase(NamespaceId.SYSTEM.getNamespace())) {
+        throw new BadRequestException("Getting specification in system namespace is currently not supported");
+      }
+
+      WorkspaceId wsId = new WorkspaceId(ns, workspaceId);
+      WorkspaceDetail detail = store.getWorkspaceDetail(wsId);
+      List<String> directives = detail.getWorkspace().getDirectives();
+      List<Row> result = executeDirectives(ns.getName(), directives, detail.getSample());
+
+      SchemaConverter schemaConvertor = new SchemaConverter();
+      Schema schema = schemaConvertor.toSchema("record", createUberRecord(result));
+      Map<String, String> properties = ImmutableMap.of("directives", String.join("\n", directives));
+
+      Set<StageSpec> srcSpecs = getSourceSpecs(detail, directives);
+
+      responder.sendString(GSON.toJson(new WorkspaceSpec(
+        srcSpecs, new StageSpec(
+          schema, new Plugin("Wrangler", "transform", properties,
+                             // TODO: find a better way to get the scope and the artifact name
+                             new Artifact("wrangler-transform", ProjectInfo.getVersion(), "system"))))));
+    });
+  }
+
+  /**
+   * Get source specs, contains some hacky way on dealing with the csv parser
+   */
+  private Set<StageSpec> getSourceSpecs(WorkspaceDetail detail, List<String> directives) {
+    SampleSpec sampleSpec = detail.getWorkspace().getSampleSpec();
+    Set<StageSpec> srcSpecs = sampleSpec == null ? Collections.emptySet() : sampleSpec.getRelatedPlugins();
+
+    // really hacky way for the parse-as-csv directive, should get removed once we have support to provide the
```

Comment:
Yes, I feel this hack should happen in the connector instead of here. When it comes to workspace stage, the sample should already have this logic applied, i.e, using blob for avro and excel file. Therefore, the generate spec should already have that.